bpf: scapy support (dev. experience)

msune opened the pull request:
This patchset adds [scapy](https://scapy.net/) support in two different ways to (hopefully) improve developer experience when writing and debugging BPF code, and when writing BPF unit tests.


Note: this patchset does NOT introduce a new testing framework.

## `hexdump()` 

`lib/hexdump.h` is a new header that defines `hexdump()` and `HEXDUMP()` primitives to `printk()` (up to) the first ~64~ 128b bytes of a packet. It is meant to be used **solely** for debugging purposes of BPF code.

`tools/log2scapy.py` can then be used to parse the `trace_pipe` output and:

* Dump the packet contents, using scapy's `.show()` routine: e.g.

```
<...>-473620  [003] b..11 955593.331365: bpf_trace_printk: /home/msuneclo/dev/cilium2/bpf/bpf_host.c:1024 l2_announce: 
###[ Ethernet ]### 
  dst       = ff:ff:ff:ff:ff:ff
  src       = de:ad:be:ef:de:ef
  type      = ARP
###[ ARP ]### 
     hwtype    = Ethernet (10Mb)
     ptype     = IPv4
     hwlen     = 6
     plen      = 4
     op        = who-has
     hwsrc     = de:ad:be:ef:de:ef
     psrc      = 110.0.11.1
     hwdst     = ff:ff:ff:ff:ff:ff
     pdst      = 172.16.10.1

```

* Open a `scapy` interactive shell with the parsed packets (e.g. to manipulate them or inject them somewhere)
* Generate a PCAP file with the parsed packets.

#### Short demo 

https://github.com/user-attachments/assets/b4db369c-a7c5-4778-bc1c-7d6ef8408971

#### Other comments:

* Never use in production code, it's just a debugging tool
* Mind complexity might increase when `hexdump()`/`HEXDUMP()`  is used

## Scapy in BPF unit tests

This patchset also adds scapy support to BPF unit tests. The support is additive, and can be used to:

* Easily generate packets using Scapy packet definitions (`BUF_DECL()`)
* Easily compare expected vs actual packet received by the unit test, and show the two if they mismatch, e.g.:

```
--- FAIL: TestBPF (12.09s)
    --- FAIL: TestBPF/tc_l2_announcement.o (0.14s)
        --- FAIL: TestBPF/tc_l2_announcement.o/1_happy_path (0.00s)
            bpf_test.go:455: tc_l2_announcement.c:156: CTX and buffer 'EXPECTED_ARP_REP' mismatch tc_l2_announcement.c:156
        bpf_test.go:239: Local container found ifindex 42 seclabel 0
FAIL
FAIL	github.com/cilium/cilium/bpf/tests/bpftest	12.110s
FAIL
=== Start tc_l2_announcement.c:156 'arp_rep_ok' ===
--- Expected ---
###[ Ethernet ]### 
  dst       = de:ad:be:ef:de:ef
  src       = 13:37:13:37:13:37
  type      = ARP
###[ ARP ]### 
     hwtype    = Ethernet (10Mb)
     ptype     = IPv4
     hwlen     = 6
     plen      = 4
     op        = is-at
     hwsrc     = 13:37:13:37:13:37
     psrc      = 172.16.10.1
     hwdst     = 13:37:13:37:13:37
     pdst      = 110.0.11.1

--- Got ---
###[ Ethernet ]### 
  dst       = de:ad:be:ef:de:ef
  src       = 13:37:13:37:13:37
  type      = ARP
###[ ARP ]### 
     hwtype    = Ethernet (10Mb)
     ptype     = IPv4
     hwlen     = 6
     plen      = 4
     op        = is-at
     hwsrc     = 13:37:13:37:13:37
     psrc      = 172.16.10.1
     hwdst     = de:ad:be:ef:de:ef
     pdst      = 110.0.11.1

===  End  tc_l2_announcement.c:156 'arp_rep_ok' ===
make: *** [Makefile:91: run] Error 1

```

You can see a test ported to scapy infra [here](https://github.com/cilium/cilium/pull/40294/commits/17f5011cf7ba2f0372b5dbcbfaa57fb3dcb1ffb7)

#### Short demo 

https://github.com/user-attachments/assets/2723888f-34a2-4bd2-8591-72d8a8de63d2

#### Other comments

* This patchset does NOT remove (nor intend) the existing pktgen infra, nor the current assert support. It is meant to be an easier way to write some tests. I expect current asserts will still be used in some cases, but most of the tests can be simplified and benefit from `scapy` syntax.
* Some TODOs / things to improve can be found [here](https://github.com/cilium/cilium/pull/40294/commits/563d42af7b1ca86d281a162520c1d6cef801c42b)
* ~Using trace_pipe for sending failed assert data back to the unit test go prog is... hacky. It might be worth looking into using a BPF map for this.~, (solved, see #43018).

## Final remarks

I'd like to get thoughts from the community on this to continue investing in it.

While we all get "used" to write BPF and BPF tests, IMO s/t like this would have saved me some precious dev time.

* ~`checkpatch.pl` errors: some style errors are a result of defining scapy pkts within BPF code. This could be sorted out by [TODO3](https://github.com/cilium/cilium/commit/4e137161a6450c26e62ae230f64cbdd9fc84a386), which would mean moving the packet definition itself to a python file (that would be loaded by `scapy/parser.py`). This might be cleaner.~ Fixed in latest commit.

Something like:

```
diff --git a/bpf/tests/tc_l2_announcement.c b/bpf/tests/tc_l2_announcement.c
index 82e3601415..4ef0570717 100644
--- a/bpf/tests/tc_l2_announcement.c
+++ b/bpf/tests/tc_l2_announcement.c
@@ -46,7 +46,7 @@ static __always_inline int build_packet(struct __ctx_buff *ctx)
 {
        struct pktgen builder;
        pktgen__init(&builder, ctx);
-       BUF_DECL(ARP_REQ, Ether(dst=mac_bcast, src=mac_one)/ARP(op="who-has", psrc=v4_ext_one, pdst=v4_svc_one, hwsrc=mac_one, hwdst=mac_bcast));
+       BUF_DECL(ARP_REQ, arp_req);
        BUILDER_PUSH_BUF(builder, ARP_REQ);
        pktgen__finish(&builder);
 
diff --git a/bpf/tests/tc_l2_announcement.py b/bpf/tests/tc_l2_announcement.py
new file mode 100644
index 0000000000..1a0e287f0d
--- /dev/null
+++ b/bpf/tests/tc_l2_announcement.py
@@ -0,0 +1 @@
+arp_req = Ether(dst=mac_bcast, src=mac_one)/ARP(op="who-has", psrc=v4_ext_one, pdst=v4_svc_one, hwsrc=mac_one, hwdst=mac_bcast)


```

* As discussed with @smagnani96, we might want to add Python lintering

msune committed:
bpf/lib: add utility ctx hexdump() helper

Add utility function that hexdumps packets via trace_printk().

WARNING: this function uses printk. It's ONLY to be used for
debugging purposes at development stage. Also, be mindful about
the complexity introduced by these helper functions/MACROs.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
bpf/tools: add log2scapy.py utility

This patch adds `log2scapy.py`, a small script that parses a trace_pipe
log (copy), and either:

* Dumps the packets along with some trace line context into stdout
* _or_ it opens a scapy interactive shell with `-i`.

The command-line option `-p <pcap_file>` optionally writes the parsed
packets in the `pcap_file` name with PCAP format.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune committed:
bpf/tests: add scapy support

This commit adds some rudimentary scapy support in BPF unit tests
to simplify the process of writing and debugging unit tests.

Note: this commit does NOT replace pktgen.h, and tests written with
pktgen.h and scapy can coexist. In fact, scapy-based tests _still_ rely
on some pktgen.h infra.

The mechanics as of today are a bit hacky, and can be broken down in
two distinct stages:

* Compilation: before clang is invoked, `scapy/parser.py` goes
  through the C BPF unit test code looking for `BUF_DECL(NAME, ...)`
  invocations. As of today, `BUF_DECL()` only accepts a single
  argument which is the scapy packet definition, e.g.:

```
  BUF_DECL(ARP_REQ, arp_req)
```

And in `bpf/tests/scapy/pkt_defs.py`:

```
arp_req = Ether(dst=mac_bcast, src=mac_one)/ARP(op="who-has", psrc=v4_ext_one, pdst=v4_svc_one, hwsrc=mac_one, hwdst=mac_bcast));
```
  (but is kept as __VA_ARGS__ as it could accept more parameters to
  be passed to the parser, see TODO2)

  `scapy/parser.py` then creates a scapy packet, and generates
  a header file `scapy/output/gen_pkts.h` with the pkt byte array. This autogen
  header is then included via `scapy.h`. As of today, the buffer is allocated
  in stack. One can reference the scapy buffer - if reachable, in scope - using
  `SCAPY_BUF(NAME)`.

  At the same time, `scapy.h` defines a MACRO to compare a pkt (ctx)
  to a scapy buffer, to simplify writing pkt unit test assertions:
  `ASSERT_CTX_BUF_OFF(...)`.

  `scapy/pkt_defs.py` contains the packet definitions to be used in the tests,
  along with global constants (e.g. `mac_one`) reused in packet definitions.
  Packet definitions should be prepended with the test name or alike.
  Symbols in there are imported (via eval!) by `parser.py`, so they are
  available when `BUF_DECL()` is parsed, hugely simplifying buffer definitions.

* Output post-processing: BPF assertions are essentially the result of
  `memcmp()`, so while writing whole pkt checks is easier, debugging
  errors is difficult. To solve this, when `ASSERT_CTX_BUF_OFF()` fails,
  it uses `hexdump()` to dump both the expected and the actual packet.
  `scapy/trace_diff_pkts.py` in combination with the MAkefile then reads
  the packets from the `trace_pipe` and outputs the 'expected' and the 'got'
  using scapy's built-in dumper (`show()`).

TODO2: support no csum computation (don't invoke show2())
TODO3: _maybe_ scope buf names to the unit test itself (right now they are global in pkt_defs.py).
TODO4: don't use `trace_pipe` and post-processing to enrich assertions.
       It might be easier to use a map to store failed assertions, and
       post-process from there.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

smagnani96 commented:
> > Agree with all TODOs:
> > * TODO1: sometimes we need to simulate L3 packets, so that's needed
>
> Actually, it might just work today with any sort of buf. The macro comparing already takes an offset, so that should work also out of the box. I will double-check.

Yep, I was referring more of when dumping the packet in `trace_diff_pkts.py` https://github.com/msune/cilium/blob/63eecfb98263fb8ce53d5b0ecf38ad10a0a84e5b/bpf/tests/scapy/trace_diff_pkts.py#L36-L37:

```py
        expected = Ether(bytes.fromhex(data["expected"]))
        got = Ether(bytes.fromhex(data["got"]))
```

> Not sure I understood your comment; you would have a separate .py file per unit test. In the example:

Ha! Ok got it. I thought you wanted to make it even more granular, like specifying the buffer only locally to the `CHECK` function rather than locally to the whole test file.

> We could could also only conditionally define HD_MAX_BYTES in hexdump.h so that you can define it before the inclusion (but then you have issues with multiple inclusions).

Or pass it as a const to the `hexdump` function, while also providing some pre-defined values (similarly as we do in the datapath for `trace_payload_len / trace_payload_len_overlay`)

borkmann commented:
[...]
> > Ha! Ok got it. I thought you wanted to make it even more granular, like specifying the buffer only locally to the CHECK function rather than locally to the whole test file.
> 
> To be honest, today you can define them in `pkt_constants.py` and it works:
> 
> ```
> # Note: these replicate pktgen.h values
> # TODO: it would be ideal to have a single source of truth for these values
> 
> ...
> 
> tcp_svc_three = 53
> 
> default_data = "Should not change!!"
> 
> # Unit tests packet definitions below!
> 
> arp_req = Ether(dst=mac_bcast, src=mac_one)/ARP(op="who-has", psrc=v4_ext_one, pdst=v4_svc_one, hwsrc=mac_one, hwdst=mac_bcast)
> ```
> 
> `scapy/parser.py` imports this file, so these symbols are available on all unit tests.
> 
> TODO3 suggests to create one of these python "envs" with the packet definitions _per unit test_ so that there are no possible conflicts. It's really a detail, because we could also continue to use a single `pkt_constants.py` (or use a better name for it), and simply prepend a unit test prefix to it, so that `bpf/tests` is not cluttered with python files. Either works for me.
> 
> > Or pass it as a const to the hexdump function, while also providing some pre-defined values (similarly as we do in the datapath for trace_payload_len / trace_payload_len_overlay)

Overall I like the series! After the base is in, I presume other devs who want to use it won't have to touch the python parts, right?

I'd think we should probably convert all the packet builders with this work to simplify the code. 

What would be great to improve imho is the macro namings, e.g. things like SCAPY_ASSERT_PKT_BUF, SCAPY_ASSERT_PKT_BUF_OFF, SCAPY_PKT_BUILDER.. maybe we can simplify the names a bit and add documentation for them with a small example.

Small nit also wrt API: SCAPY_ASSERT_PKT_BUF_OFF has `CTX` as an arg, in the do {} while body we also use `ctx`. As I understand both are the same, but in some cases we use CTX and in others ctx? This feels prone to unexpected bugs, rather use `(CTX)` everywhere.

smagnani96 commented:
Another side note -- no need to address in the first PR.
Sometimes in tests we do modify some `ctx` layer info within the same `CHECK` instead of writing a whole new test `PKTGEN+CHECK`. An example could be found [here](https://github.com/cilium/cilium/blob/main/bpf/tests/classifiers_common.h#L153).
I'm wondering if it can make sense to add helpers to manipulate the `ctx` data on-the-fly, to avoid having to unpack and check pointers in C to modify some field (ex L4 ports).

msune committed:
bpf/tests: tc_l2_announcement.c use scapy

This commit refactors `tc_l2_announcement.c` to make use of the new
scapy test infrastructure (showcase).

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

Artyop commented on .github/workflows/lint-bpf-checks.yaml:
```diff
@@ -119,6 +119,11 @@ jobs:
       job_name: "bpf_tests"
     runs-on: ubuntu-24.04
     steps:
+      - name: Install Scapy (and Python3)
```
I'd suggest setup python with [gha](https://github.com/actions/setup-python) for version tracking and install scapy using pip on a second step

msune replied:
Done here: https://github.com/cilium/cilium/pull/40294/commits/4f4465747f786575553c0a00d1f48e230a95e573

msune replied:
> If BPF unit tests are running under cilium-builder container, then we're talking about this one https://github.com/cilium/cilium/tree/main/images/builder

Agreed.

> I feel like overriding https://github.com/cilium/cilium/blob/main/images/builder/install-builder-deps.sh with python install and pip install scapy should be a quick and clean way of handling this

Yup, I will open a PR. In the meantime I will create a tc commit on top that install scapy in that container during run to start running the tests in the CI

> (and also I think you'd need to use pip and not pip3 after setting up python with the action)

`pip3` was working, but I had to symlink `python3` to `python`. That seems a bit stupid from that GHA... as python3 should also be in the PATH.

msune replied:
Fixed https://github.com/cilium/cilium/pull/40838

msune commented on bpf/tests/Makefile:
```diff
@@ -23,11 +23,26 @@ else
+all: $(TEST_OBJECTS)
+
+$(SCAPY_HDR):
+	$(QUIET) python3 scapy/parser.py . > $@.tmp
```
Yes. It should, else tests can't run.

What is the concern?

smagnani96 replied:
1. Your [above comment](https://github.com/cilium/cilium/pull/40294#discussion_r2229327895) for the builder image 
2. I did not remember we already [cite in the doc](https://docs.cilium.io/en/latest/contributing/development/dev_setup/#:~:text=python3%2Dpip,A%20(OS%2Dspecific)) to have python3 (also in `make dev-doctor`) for dev

msune replied:
Folowing in the other thread

msune commented on bpf/tests/scapy/parser.py:
```diff
@@ -0,0 +1,86 @@
+import os
```
I will do that and attribute to you & me.

msune replied:
On:

```
#!/usr/bin/env python3
```

I am quite neutral on invoking this scripts as `python3 script.py` vs `./script.py`. If you prefer that, I can do that

msune replied:
Both fixed.

msune committed:
bpf: add LICENSE notice in Python files

Add SPDX Apache license identifier in scapy-related Python scripts.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>
Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

tommyp1ckles commented on bpf/tests/scapy/parser.py:
```diff
@@ -0,0 +1,91 @@
+#!/usr/bin/env python3
```
Just a thought; although scapy is a python based framework, writing this tooling in Python seems we're introducing another language to the toolchain used for building/generating/running bpf tests.  Would it make sense to have the parsing/templating tooling portion be written in Go and perhaps be part of the existing bpf test runner infra Code (i.e. while just shelling out to the scapy CLI to do the byte array generation).
(This would probably also make it much easier to write the code that would read from a map for the diff output as you mention in your TODOs)

This might ultimately be more complex, just wanted to bring up the topic..

smagnani96 replied:
> Besides, the project already has other Python scripts e.g. verifier_diff.py or consolidate_go_stacktrace.py, doc ones etc.

WFIW, here we're discussing about a script that'd be invoked during the build chain. The ones mentioned here should be completely out of that loop.

A compromise between the what Tom suggested and the PR would be to keep the separate Python files, but having our GO bpf test runner setup the infra (trace pipe) and calling the `.py` scripts. Thoughts?

msune replied:
> but having our GO bpf test runner setup the infra (trace pipe) and calling the .py scripts. 

Possible? yes. But what are the pros/cons?

* I looked into it and there was more code than natively with Python. I happy to be proven wrong here.
* Performance boost: likely negligible, except perhaps on the `parser.py` (and yet to be seen).
* The build chain _ends up calling Python anyhow_, no matter it is via Go
* The Python code - the same - needs to continue to be maintained. Python scripts are very very simple. They don't do anything special, it's mostly calling Scapy and 2 regexes. As said before, the only one that _might_ be worth the effort is `parser.py`, but still we will need to continue maintaining Python code.

As I was mentioning to @dylandreimerink, I think we need to use the right tool for the right job. I totally understand the resistance to Python, but this is not something that will "spread" within the codebase. We might use scapy for functional testing in other parts, but it's very much limited to that. I've personally yet to see s/t that works better for function testing dataplanes (complex pkt generation, comparison, checksum etc..), and I've used HW/SW traffic generators (some very expensive and fancy) and a lot of similar tooling in past lives.

tommyp1ckles commented on bpf/tests/scapy/parser.py:
```diff
@@ -0,0 +1,91 @@
+
+# Match SCAPY_BUF(name, ...
+PKT_REGEX = re.compile(
+    r'SCAPY_DEF_BUF\(\s*([A-Za-z0-9_]+)\s*,\s*(.+?)\s*$',
```
This might cause some awkward cases, like if someone writes a comment regarding some usage of `SCAPY_DEF_BUF(...)` that might accidentally get picked up here and expanded.

msune replied:
Good catch,

I will start the regex with `^'\s*` to match newline + any space, tab etc to avoid this. With the style of comments that we support (`checkpatch.pl`), that should be sufficient.

msune commented on bpf/lib/hexdump.h:
```diff
@@ -49,8 +49,8 @@ void hexdump_len_off(const char *msg, const char *first_layer,
-	__u32 pkt_len = (data_end - data);
-	__u32 max_dump_len = len > HD_MAX_BYTES ? HD_MAX_BYTES : len;
+	__u16 pkt_len = (__u16)(data_end - data);
+	__u16 max_dump_len = len > HD_MAX_BYTES ? HD_MAX_BYTES : len;
```
self review: move to hexdump commit

msune commented on bpf/tests/scapy/pkt_defs.py:
```diff
@@ -0,0 +1,64 @@
+
+# IPv4 addresses for hosts, external to the cluster
+v4_ext_one   = "110.0.11.1"
+v4_ext_two	 = "120.0.12.2"
```
self review: tab

msune commented on bpf/.gitignore:
```diff
@@ -3,3 +3,5 @@ compile_commands.json
 *.i
 *.s
 .rebuild_all
+tests/scapy/*
```
Done

msune committed:
builder.sh: mount /sys for BPF unit tests (scapy)

This commit mounts /sys when invoking contrib/scripts/builder.sh,
so that BPF unit tests with the scapy support can read from the
trace_pipe.

It also changes builder.sh to receive quoted DOCKER_ARGS.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

bowei commented on bpf/tools/log2scapy.py:
```diff
@@ -0,0 +1,65 @@
+def parse_pkts(filename: str) -> List[Dict]:
+    #TODO attempt to parse FILENAME:LINENUM if possible (HEXDUMP())
+    #TODO parse timestamp
+    pattern = re.compile(r"\s*(.*)\s*pkt_hex\s*(\w+)\[(.*?)\]")
```
nit: might want to put the format and an example in the doc string (regex tends to get somewhat hard to decipher over time).

msune replied:
Added [here](https://github.com/cilium/cilium/pull/40294/commits/8ef304b48d499a350e7d8bffcc8907af333874f0#diff-68716399c65f043968da9c5339c8520fb00c7e883ee7c3fd2836e15bcdd1e58bR12-R19) and [here](https://github.com/cilium/cilium/pull/40294/commits/430bb1dea75066172f0b73ca7c9d983aa559c280#diff-a2e7f22890560c0d9bfcc7983e890dfafcd5c22b9d08a232ba363ddd3a537848R13-R23).

msune commented on Makefile:
Dependency PR opened: https://github.com/cilium/cilium/pull/40838

qmonnet commented on contrib/scripts/builder.sh:
```diff
@@ -54,6 +54,7 @@ docker run --rm \
 	$MOUNT_GOCACHE_DIR \
 	$MOUNT_GOMODCACHE_DIR \
 	$MOUNT_CCACHE_DIR \
+	-v /sys:/sys \
```
Why is this necessary?

msune replied:
To be able to access the trace_pipe to recover the pkt dumps:

* https://github.com/cilium/cilium/pull/40294/commits/563d42af7b1ca86d281a162520c1d6cef801c42b#diff-d842ee12721dfe61b336fa06a6646dd78f2f733e8000c2d014fc01ff2cd5c0a8R103

Now that you mention it, it might be a good idea to move it to `DOCKER_ARGS` in the `run_bpf_tests` `Makefile`'s target

qmonnet reviewed (commented):
Looks great!

> Never use [`hexdump()`/`HEXDUMP()`] in production code, it's just a debugging tool

Would it make sense to have a check (coccicheck?) to make sure that these macro never make it to the datapath code?

msune commented:
> Assuming the To be removed, ... commits are removed before merging.

:100: Once #40838 is in, I will remove all "To be removed" patches.

>@qmonnet wrote:
>> Never use [hexdump()/HEXDUMP()] in production code, it's just a debugging tool
>Would it make sense to have a check (coccicheck?) to make sure that these macro never make it to the datapath code?

I think this is a great idea. Let me add a patch on top. I will ping you to review it if you don't mind. I will also action the `-v /sys:/sys` thing.

msune committed:
bpf/test/scapy: add scapy documentation

Add a small guide on how to write BPF unit tests with Scapy along
with a short intro to Scapy.

Signed-off-by: Marc Suñé <marc.sune@isovalent.com>

msune commented:
> @dylandreimerink wrote:
>Also, could we include a link to scapy docs somewhere?. I was unable to find a good docs/references page to find out what the struct/object names and fields/arguments are that are available.

I went the extra mile and created a small guide :book:  [here](https://github.com/cilium/cilium/pull/40294/commits/9e89f8543db016f7a25c5d721711b64efc56e2b4)

> @qmonnet wrote:
> Would it make sense to have a check (coccicheck?) to make sure that these macro never make it to the datapath code?

I am new to coccicheck and every attempt I have tried so far failed miserably. See the next paragraph.

---

The two things I am missing:

* coccicheck: maybe can we add it in a separate PR @qmonnet?
* Doing a final round of "failed" BPF unit tests to see that everything is working well, to then remove the "To remove" commits.  On this, the cilium-builder image doesn't seem to have python3+scapy+jinja2 after https://github.com/cilium/cilium/pull/40838 :thinking:. Any thoughts @HadrienPatte?

```
~/dev/cilium$ docker run -it quay.io/cilium/cilium-builder:769ce739568d824078d2cf261de649c54ecb1315@sha256:413740790b2a15ff889c4f91ac2a2dda847159f7778acf1523a196d0edbedf12 bash
root@ef79392872a3:/go/src/github.com/cilium/cilium# python3
bash: python3: command not found
root@ef79392872a3:/go/src/github.com/cilium/cilium# scapy
bash: scapy: command not found
```

msune commented:
>I see you updated images/builder/install-builder-deps.sh. But is this file actually used at all? I can't find how or where it would be called (including from the commit that introduced it 🙀). Isn't images/builder/Dockerfile the one containing the actual dependencies?

Ok, let's try it again, now updating the `Dockerfile` :smile:

smagnani96 commented:
> @smagnani96 I see you self-requested a review a few days ago, do you want to take another look or are you good with this PR?

I'm reviewing this right now, was waiting for the force-pushes 🙏🏼

smagnani96 reviewed (approved):
LGTM! Would probably create (if not done already) an GH issue to track possible TODOs and enhancements to this (e.g., the cocci script).

Amazing Marc! 🚀

msune commented on Makefile:
```diff
@@ -546,7 +546,7 @@ BPF_TEST_DUMP_CTX ?= ""
 
 run_bpf_tests: ## Build and run the BPF unit tests using the cilium-builder container image.
-	DOCKER_ARGS=--privileged RUN_AS_ROOT=1 contrib/scripts/builder.sh \
+	DOCKER_ARGS="--privileged -v /sys:/sys" RUN_AS_ROOT=1 contrib/scripts/builder.sh \
```
I kind of lost track on the reasoning, but IIRC the problem was `DOCKER_ARGS` was generated from the targets invoking the builder [script unquoted](https://github.com/cilium/cilium/blob/f2239b71bae3affe59edac7ee005659b448ce710/Makefile#L549). That worked fine for a single arg (e.g. `--privileged` for `run_bpf_tests`) but failed for more than one arg.

The solution was to obviously quote the `DOCKER_ARGS` in the original `Makefile`, but then the issue was that https://github.com/cilium/cilium/blob/f2239b71bae3affe59edac7ee005659b448ce710/contrib/scripts/builder.sh#L59 was double quoting and hence failing.

qmonnet replied:
https://github.com/cilium/cilium/pull/40294#discussion_r2244985973

msune commented on bpf/.gitignore:
```diff
@@ -6,3 +6,7 @@ compile_commands.json
 
 # ignore CLANG intermediate compilation tmp files
 *.o.tmp
+
```
ACK, let's see.
